### PR TITLE
Add back LicenseManager.LicenseInteropHelper

### PR DESCRIPTION
**Pull request by ericstj**
Fixes https://github.com/dotnet/core/issues/1994 

Essentially undoes https://github.com/dotnet/corefx/commit/d067d8de57d817fe7e9f2e17cd9d776b187502f3 and ensures that this code is not trimmed.

@erozenfeld I found that listing the nested type name as `System.ComponentModel.LicenseManager.LicenseInteropHelper` wasn't sufficient to root it, but rooting the GetLicenseInteropHelperType seems to root the entire type.  Can I count on that moving forward?

### CODE REVIEW

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/LicenseManager.cs [hunk 1]
```diff
@@ -357,5 +357,12 @@ public static License Validate(Type type, object instance)
 
             return lic;
         }
+
+        /// <summary>
+        /// Retrieves the typehandle of the interop helper
+        /// </summary>
+        private static IntPtr GetLicenseInteropHelperType() {
```
Comment (jkotas) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/LicenseManager.cs]: We should not have this method. It is useless indirection. The CoreCLR should be fixed to fetch LicenseInteropHelper directly.

**Comment by jkotas** [comment]
We need to have a test for this.

**Comment by erozenfeld** [comment]
Nested type name should be listed as
```
System.ComponentModel.LicenseManager/LicenseInteropHelper
```

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/LicenseManager.LicenseInteropHelper.cs [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+            private const int S_OK = 0;
+            private const int E_NOTIMPL = unchecked((int)0x80004001);
+            private const int CLASS_E_NOTLICENSED = unchecked((int)0x80040112);
+            private const int E_FAIL = unchecked((int)0x80000008);
+            private DesigntimeLicenseContext _helperContext;
+            private LicenseContext _savedLicenseContext;
+            private Type _savedType;
+            // The CLR invokes this whenever a COM client invokes
+            // IClassFactory::CreateInstance() or IClassFactory2::CreateInstanceLic()
+            // on a managed managed that has a LicenseProvider custom attribute.
```
Comment (stephentoub) [inline src/System.ComponentModel.TypeConverter/src/System/ComponentModel/LicenseManager.LicenseInteropHelper.cs]: managed managed?

**Comment by ericstj** [comment]
> We need to have a test for this.

Sure,  I was considering this and didn't per your [comment](https://github.com/dotnet/core/issues/1994#issuecomment-429466197) about @AaronRobinsonMSFT improving testing.

Since this is a private type are folks ok with a test that does private reflection?

I'll also see if I can trigger the interop scenario with a minimal implementation of something like the original issue only using CoreFx code.

File: src/System.ComponentModel.TypeConverter/src/System/ComponentModel/LicenseManager.cs (hunk 1 above)
Comment (AaronRobinsonMSFT) [reply src/System.ComponentModel.TypeConverter/src/System/ComponentModel/LicenseManager.cs]: Also, it should be called out that my intent for COM activation is that it should be done in managed as much as possible - this code will be definitely be consumed from SPCL for managed COM server activation. See https://github.com/dotnet/coreclr/blob/ffc347acdbcaa05bc85fcd51bca57c551bc33955/src/System.Private.CoreLib/src/System/Runtime/InteropServices/ComActivator.cs#L240

**Comment by jkotas** [comment]
> Since this is a private type are folks ok with a test that does private reflection?

The test should be testing the actual scenario that did not work before. The test should be using private reflection - there is not much value in testing the internal implementation details.

**Comment by ericstj** [comment]
Got it I'll focus on the scenario testing.

My suggestion for the reflection-based testing was really around the fact that this private API is treated as visible by other assemblies so it might be useful to treat it like public and test it that way, but that's probably wasted effort giving your suggestions to change how this works.

**Comment by ericstj** [comment]
I tried to trick interop into calling this code from managed code (without actually creating a native com server / client) and didn't have luck.  I've filed https://github.com/dotnet/corefx/issues/32999 to track improving the test coverage.

I added a simple regression test for the broken scenario and constrained it to windows.  Made sure it AV'ed prior to this change and succeeds with the fix in place.

**Comment by ericstj** [comment]
test Linux x64 Release Build

**Comment by ericstj** [comment]
test Linux x64 Release Build

**Comment by ericstj** [comment]
test Linux x64 Release Build

**Pull request merged**